### PR TITLE
SoftDeterministicPolicyNetwork: fixed inconsistent return types in training and eval mode.

### PR DESCRIPTION
**Pull request by michalgregor**
I have discovered a slight bug on develop, regarding SAC in eval mode. The SoftDeterministicPolicyNetwork's forward returns a tensor in eval mode and a tuple in training mode. The agent then gets the 0-th dimension of the returned value in both cases – so in eval mode, SAC fails, returning a scalar instead of a vector as action.

Also, this used to work before accidentally, because the network returned a 2D tensor: so in training mode, the [0] would have selected that tensor from the tuple and in eval mode it would have selected the first row from the 2D tensor (which would still work, even though it was not really intended).

Finally, the pull request is showing me 3 commits instead of 1. I don't know why that is precisely, since the first 2 commits are already on develop as far as I can see. If you know why this is happening, please, let me know and I will fix it before you merge.

### CODE REVIEW

File: all/policies/soft_deterministic.py [hunk 1]
```diff
@@ -43,7 +43,7 @@ def forward(self, state):
         if self.training:
             action, log_prob = self._sample(normal)
             return action, log_prob
-        return self._squash(normal.loc)
+        return self._squash(normal.loc), torch.as_tensor(0.0, device=self.device)
```
Comment (cpnota) [inline all/policies/soft_deterministic.py:46]: Is this the right fix, or would the right fix be to change `SAC`? The log_prob isn't really 0.0, per se. A third option would be to actually compute the log_prob for the greedy action? Still sort of misleading, perhaps.

File: all/presets/atari/a2c.py [hunk 2]
```diff
@@ -23,6 +23,10 @@ def a2c(
         # Batch settings
         n_envs=16,
         n_steps=5,
+        # Model construction
+        feature_model_constructor=nature_features,
+        value_model_constructor=nature_value_head,
+        policy_model_constructor=nature_policy_head
```
Comment (cpnota) [inline all/presets/atari/a2c.py:29]: Not sure with GitHub is showing these as changes, looks okay in the file view, though.

File: all/environments/gym.py [hunk 3]
```diff
@@ -6,9 +6,13 @@
...
 class GymEnvironment(Environment):
     def __init__(self, env, device=torch.device('cpu')):
-        self._name = env
+        
         if isinstance(env, str):
+            self._name = env
             env = gym.make(env)
+        else:
+            self._name = env.__class__.__name__
+
```
Comment (cpnota) [inline all/environments/gym.py:15]: Nice catch!

File: all/policies/soft_deterministic.py (hunk 1 above)
Comment (michalgregor) [reply all/policies/soft_deterministic.py:46]: I can totally change that, but my reasoning was this: unless I am misreading the code, the policy is deterministic in eval mode, so the probability of the selected action is really 1, right? So its log would then be 0?

File: all/environments/gym.py (hunk 3 above)
Comment (michalgregor) [reply all/environments/gym.py:15]: Yeah, I wanted to do this in a separate PR originally, but apparently I am missing something crucial as to how PRs work here on GitHub. :D :D I will need to look into that...

**Comment by cpnota** [comment]
Ended up going with a slightly different fix. The fixes were handled by #169 and #170 . Thanks again for identifying these bugs!
